pydantic/_internal/_fields.py

```diff
@@ -53,6 +63,52 @@ def get_type_hints_infer_globalns(
     return get_type_hints(obj, globalns=globalns, localns=localns, include_extras=include_extras)
 
 
+def get_typed_dict_type_hints(
+    typed_dict_cls: Any,
+    types_namespace: dict[str, Any] | None = None,
+) -> dict[str, Any]:
+    """Collect the type hints of a TypedDict.
+
+    This takes into account any TypeVar transformations when using generic
+    TypedDict.
+
+    Args:
+        typed_dict_cls: a TypedDict subclass
+        types_namespace: Optional extra namespace to look for types in.
+
+    Returns:
+        the final type hints.
+    """
+    hints: dict[str, Any] = {}
+    typevars_map: dict[type, dict[Any, Any]] = {}
+    previous_base: type | None = None
+
+    for base_cls in reversed(mro(typed_dict_cls)):
+        base = base_cls
+        if base in typed_dict_types:
+            continue
+
+        # needed if base_cls is a generic alias
+        if get_args(base_cls):
+            base: Any = get_origin(base_cls)
+
+        new_type_hints = eval_cls_annotations_lenient(base, types_namespace)
+
+        typevars_map[base] = get_standard_typevars_map(base_cls) or {}
+
+        for field_name, annotation in new_type_hints.items():
+            if field_name in hints:
+                if previous_base is not None:
+                    hints[field_name] = replace_types(hints[field_name], typevars_map[previous_base])
+                continue
```

Comment:
For what it's worth, it looks like this `if` / `continue` pair is ok because it's not legal to specify the same typed dict field more than once through the whole inheritance tree, so the type cannot be refined by subclasses. Perhaps worth adding a comment?

https://peps.python.org/pep-0589/#inheritance